File: collect_app/src/main/java/org/odk/collect/android/views/TrackingTouchSlider.java
package org.odk.collect.android.views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.slider.Slider;

import org.jetbrains.annotations.NotNull;

public class TrackingTouchSlider extends Slider implements Slider.OnSliderTouchListener {

    private boolean trackingTouch;
    private boolean enabled;

    private OnMinValueChangedListener onMinValueChangedListener;

    public interface OnMinValueChangedListener {
        void onFirstValueChanged();
    }

    @SuppressLint("ClickableViewAccessibility")
    public TrackingTouchSlider(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        addOnSliderTouchListener(this);
        setLabelFormatter(null);
        setOnTouchListener((v, event) -> {
            int action = event.getAction();
            switch (action) {
                case MotionEvent.ACTION_DOWN:
                    v.getParent().requestDisallowInterceptTouchEvent(true);
                    break;
                case MotionEvent.ACTION_UP:
                    v.getParent().requestDisallowInterceptTouchEvent(false);
                    if (!enabled) {
                        onMinValueChangedListener.onFirstValueChanged();
                    }
                    break;
            }
            v.onTouchEvent(event);
            return true;
        });
    }

    public boolean isTrackingTouch() {
        return trackingTouch;
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void onStopTrackingTouch(@NotNull Slider slider) {
        trackingTouch = false;
    }

    @SuppressLint("RestrictedApi")
    @Override
    public void onStartTrackingTouch(@NotNull Slider slider) {
        trackingTouch = true;
    }

    public void enable() {
        enabled = true;
    }

    public void disable() {
        enabled = false;
    }

    public void setOnFirstValueChanged(OnMinValueChangedListener onMinValueChangedListener) {
        this.onMinValueChangedListener = onMinValueChangedListener;
    }
}

File: collect_app/src/main/java/org/odk/collect/android/views/TrackingTouchSlider.kt
package org.odk.collect.android.views

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.util.AttributeSet
import com.google.android.material.R
import com.google.android.material.color.MaterialColors
import com.google.android.material.slider.Slider

@SuppressLint("ClickableViewAccessibility")
class TrackingTouchSlider(
    context: Context,
    attrs: AttributeSet?
) : Slider(context, attrs), Slider.OnSliderTouchListener {
    private var listener: OnChangeListener? = null

    private var defaultTickActiveTintList: ColorStateList = tickActiveTintList
    private var defaultThumbWidth = thumbWidth
    private var defaultThumbTrackGapSize = thumbTrackGapSize

    var isTrackingTouch: Boolean = false
        private set

    init {
        addOnSliderTouchListener(this)
        setLabelFormatter(null)
        addOnSliderTouchListener(object : OnSliderTouchListener {
            override fun onStartTrackingTouch(slider: Slider) {}

            override fun onStopTrackingTouch(slider: Slider) {
                setTickActiveTintList(defaultTickActiveTintList)
                setThumbWidth(defaultThumbWidth)
                setThumbTrackGapSize(defaultThumbTrackGapSize)
                listener?.onValueChange(this@TrackingTouchSlider, value, true)
            }
        })
    }

    @SuppressLint("RestrictedApi")
    override fun onStopTrackingTouch(slider: Slider) {
        isTrackingTouch = false
    }

    @SuppressLint("RestrictedApi")
    override fun onStartTrackingTouch(slider: Slider) {
        isTrackingTouch = true
    }

    fun reset() {
        value = valueFrom
        setTickActiveTintList(
            ColorStateList.valueOf(
                MaterialColors.getColor(
                    this,
                    R.attr.colorPrimary
                )
            )
        )
        setThumbWidth(0)
        setThumbTrackGapSize(0)
    }

    fun setListener(listener: OnChangeListener) {
        this.listener = listener
    }
}
